crates/ra_assists/src/merge_match_arms.rs

```diff
@@ -0,0 +1,189 @@
+use hir::db::HirDatabase;
+use ra_syntax::{
+    ast::{AstNode, MatchArm},
+    algo::non_trivia_sibling, Direction, T
+};
+use crate::{Assist, AssistCtx, AssistId};
+
+pub(crate) fn merge_match_arms(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let comma = ctx.token_at_offset().find(|leaf| leaf.kind() == T![,])?;
+    let prev = non_trivia_sibling(comma.clone().into(), Direction::Prev)?;
+    let next = non_trivia_sibling(comma.clone().into(), Direction::Next)?;
+
+    if let (Some(p), Some(n)) = (prev.as_node(), next.as_node()) {
+        if let (Some(p), Some(n)) = (MatchArm::cast(p.clone()), MatchArm::cast(n.clone())) {
+            // Don't try to handle arms with guards for now - can add support for this later
+            if p.guard().is_some() || n.guard().is_some() {
+                return None;
+            }
+
+            let (p_expr, n_expr) = (p.expr()?, n.expr()?);
+
+            // Check for match arm equality by comparing lengths and then string contents
+            if p_expr.syntax().text_range().len() != n_expr.syntax().text_range().len() {
+                return None;
+            }
+
+            let p_text = p_expr.syntax().text().to_string();
+            let n_text = n_expr.syntax().text().to_string();
+
+            if p_text != n_text {
+                return None;
+            }
+
+            ctx.add_action(AssistId("merge_match_arms"), "merge match arms", |edit| {
+                fn contains_placeholder_pattern(a: &MatchArm) -> bool {
+                    a.pats().any(|x| match x.kind() {
+                        ra_syntax::ast::PatKind::PlaceholderPat(..) => true,
+                        _ => false
+                    })
+                }
+
+                let pats = if contains_placeholder_pattern(&p) || contains_placeholder_pattern(&n) {
+                    "_".into()
+                } else {
+                    let ps: Vec<String> = p.pats().map(|x| x.syntax().to_string())
+                        .chain(
+                            n.pats().map(|x| x.syntax().to_string())
+                        ).collect();
+                    ps.join(" | ")
+                };
+
+                let arm = format!("{} => {}", pats, p_text);
+
```

Comment:
What should I do here? I've failed to find the methods to let me do what I want.

Reply:
I think you can delete the text from `p.syntax().text_range().start()` to `n.syntax().text_range().end()`. This should take care of commas as well.

Reply:
Or, better yet, just `TextEditBuilder::replace` this range